README.md
# MatomoWire
ProcessWire module to embed Matomo (aka Piwik). Also integrates with [PrivacyWire](https://github.com/blaueQuelle/privacywire/).
* [Module Repo](https://processwire.com/modules/matomo-wire/)
* [Support Forum](https://processwire.com/talk/topic/24685-matomowire-embedding-matomo)

## How to use
After installing this module you can configure most of the options in the MatomoWire module config. The settings belonging the Textformatter (to provide Opt-Out possibilities) are within the "MatomoWire Textformatter" module configs.

This module requires an installed Matomo instance. Insert your Matomo URL and the Site ID or Tag Manager Container ID of the current website. 

## Changelog
see [CHANGELOG.md](CHANGELOG.md)


CHANGELOG.md
# Changelog of MatomoWire

## 0.0.5
- Merged PR from @horst-n: Disable PrivacyWire related settings, if PrivacyWire is not installed
- Optimized Output, if no Matomo URL is provided

## 0.0.4
- Added: additional check before rendering 

## 0.0.3 Launch
- Fixed: Typo in README.md

## 0.0.2
- Updated Readme

## 0.0.1 Init
- Initial start of MatomoWire

TextformatterMatomoWireConfig.php
<?php namespace ProcessWire;

class TextformatterMatomoWireConfig extends ModuleConfig
{
    public function getDefaults()
    {
        return [
            'open_tag' => '[[',
            'close_tag' => ']]',
            'tag_name' => "matomowire-opt-out"
        ];
    }

    public function getInputfields()
    {
        $inputfields = parent::getInputfields();

        // open tag
        $f = $this->modules->get('InputfieldText');
        $f->attr('name', 'open_tag');
        $f->label = $this->_('Open Tag');
        $f->columnWidth = 33;
        $inputfields->add($f);

        // close tag
        $f = $this->modules->get('InputfieldText');
        $f->attr('name', 'close_tag');
        $f->label = $this->_('Close Tag');
        $f->columnWidth = 33;
        $inputfields->add($f);

        // tag name
        $f = $this->modules->get('InputfieldText');
        $f->attr('name', 'tag_name');
        $f->label = $this->_('Tag name');
        $f->columnWidth = 34;
        $inputfields->add($f);

        return $inputfields;
    }
}


MatomoWireConfig.php
<?php namespace ProcessWire;

class MatomoWireConfig extends ModuleConfig
{
    public function getDefaults()
    {
        return [
            'matomo_url' => '',
            'embedding_method' => 'site',
            'render_position' => 'body',
            'site_id' => '',
            'container_id' => '',
            'set_do_not_track' => false,
            'disable_cookies' => false,
            'integrate_privacywire' => false,
            'privacywire_cookie_category' => 'statistics',
            'opt_out_type' => 'iframe'
        ];
    }

    public function getInputfields()
    {
        $inputfields = parent::getInputfields();


        // Matomo URL
        $f = $this->modules->get('InputfieldURL');
        $f->attr('name', 'matomo_url');
        $f->label = $this->_('Matomo URL');
        $f->description = $this->_('Insert the full URL including http:// or https://');
        $f->columnWidth = 33;
        $inputfields->add($f);

        // Embedding Method
        $f = $this->modules->get('InputfieldSelect');
        $f->attr('name', 'embedding_method');
        $f->description = $this->_("Choose between the classical matomo site tracking or via Matomo Tag Manager");
        $f->label = $this->_('Embedding Method');
        $f->options = [
            "site" => $this->_("Site Tracking"),
            "tag" => $this->_("Matomo Tag Manager Tracking"),
        ];
        $f->columnWidth = 33;
        $f->required = 1;
        $inputfields->add($f);

        // Rendering Position
        $f = $this->modules->get('InputfieldSelect');
        $f->attr('name', 'render_position');
        $f->description = $this->_("Choose between rendering within <head> or within <body>.");
        $f->label = $this->_('Rendering Position');
        $f->options = [
            "head" => $this->_("<head>"),
            "body" => $this->_("<body>"),
        ];
        $f->columnWidth = 34;
        $inputfields->add($f);

        // Site Id
        $f = $this->modules->get('InputfieldText');
        $f->attr('name', 'site_id');
        $f->label = $this->_('Site Id');
        $f->columnWidth = 100;
        $f->showIf("embedding_method=site");
        $inputfields->add($f);

        // Container Id
        $f = $this->modules->get('InputfieldText');
        $f->attr('name', 'container_id');
        $f->label = $this->_('Container Id');
        $f->columnWidth = 100;
        $f->showIf("embedding_method=tag");
        $inputfields->add($f);

        // Site Tracking: Respect DoNotTrack (DNT)
        $f = $this->modules->get('InputfieldCheckbox');
        $f->attr('name', 'set_do_not_track');
        $f->showIf("embedding_method=site");
        $f->label = $this->_('Respect Do-Not-Track');
        $f->columnWidth = 50;
        $inputfields->add($f);

        // Site Tracking: Disable Cookies
        $f = $this->modules->get('InputfieldCheckbox');
        $f->attr('name', 'disable_cookies');
        $f->description = $this->_('It is possible to track users [without cookies](https://matomo.org/faq/new-to-piwik/how-do-i-use-matomo-analytics-without-consent-or-cookie-banner/) - [the data accuracy is impacted](https://matomo.org/faq/general/faq_156/)');
        $f->showIf("embedding_method=site");
        $f->label = $this->_('Disable Cookies');
        $f->columnWidth = 50;
        $inputfields->add($f);

        // Fieldset for PrivacyWire Integration
        $privacywireIsInstalled = $this->modules->isInstalled('PrivacyWire');
        $privacywire_integration = $this->modules->get('InputfieldFieldset');
        $privacywire_integration->label = $this->_('PrivacyWire Integration');
        if(!$privacywireIsInstalled) {
            $privacywire_integration->collapsed = Inputfield::collapsedNoLocked;  // Inputfield::collapsedLocked
            $privacywire_integration->description = $this->_('**Important:** PrivacyWire needs to be installed to enable this feature!') . ' ([https://github.com/blaueQuelle/privacywire](https://github.com/blaueQuelle/privacywire))';
        }
        $inputfields->add($privacywire_integration);

        // Enable PrivacyWire Integration
        $f = $this->modules->get('InputfieldCheckbox');
        $f->attr('name', 'integrate_privacywire');
        $f->label = $this->_('Enable PrivacyWire Integration');
        $f->columnWidth = 33;
        if(!$privacywireIsInstalled) $f->collapsed = Inputfield::collapsedLocked;
        $privacywire_integration->add($f);

        // PrivacyWire Cookie Category
        $f = $this->modules->get('InputfieldSelect');
        $f->attr('name', 'privacywire_cookie_category');
        $f->showIf("integrate_privacywire=1");
        $f->description = $this->_("Choose, which PrivacyWire Cookie Category should trigger the Matomo Integration");
        $f->label = $this->_('PrivacyWire Cookie Category');
        $f->options = [
            "necessary" => $this->_("Necessary Cookies"),
            "functional" => $this->_("Functional Cookies"),
            "all" => $this->_("All Cookies"),
            "statistics" => $this->_("Statistics"),
            'marketing' => $this->_("Marketing"),
            'external_media' => $this->_("External Media")
        ];
        $f->columnWidth = 33;
        if(!$privacywireIsInstalled) $f->collapsed = Inputfield::collapsedLocked;
        $privacywire_integration->add($f);

        // PrivacyWire Cookie Category
        $f = $this->modules->get('InputfieldSelect');
        $f->attr('name', 'opt_out_type');
        $f->showIf("integrate_privacywire=1");
        $f->label = $this->_('Matomo Opt-Out Type');
        $f->description = $this->_("Should the Opt-Out Tag (Textformatter) trigger the Matomo iFrame Opt-Out (default) or the PrivacyWire Choose-Cookies-Window?");
        $f->options = [
            "iframe" => $this->_("Matomo iFrame (default)"),
            "privacywire" => $this->_("PrivacyWire Choose-Cookies-Window"),
        ];
        $f->columnWidth = 34;
        if(!$privacywireIsInstalled) $f->collapsed = Inputfield::collapsedLocked;
        $privacywire_integration->add($f);

        return $inputfields;
    }
}
